### PR TITLE
[DEVHUB-1424] Add On-Demand Revalidation

### PR DESCRIPTION
**Pull request by mcolella14**
[DEVHUB-1424](https://jira.mongodb.org/browse/DEVHUB-1424)

https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation

In conjunction with https://github.com/10gen/devhub-cms/pull/146, when a content piece is updated in the CMS, a POST will be sent to `/academia/api/revalidate`, and will rebuild both the home page, and that content piece's corresponding page (if it is a course).

### CODE REVIEW

File: src/lib/cms-content.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ const itemMap = ({
   return {
     id,
     title,
-    contentType,
+    contentType: contentType === 'CaseStudy' ? 'Case Study' : contentType,
```
Comment (mcolella14) [inline src/lib/cms-content.ts:40]: In the CMS, we can't have spaces in enum fields, so this comes from the api as `CaseStudy` now.

File: src/pages/courses/[course].tsx [hunk 2]
```diff
@@ -54,7 +50,7 @@ export const getStaticPaths: GetStaticPaths = async () => {
 
   return {
     paths,
-    fallback: false,
+    fallback: 'blocking',
```
Comment (mcolella14) [inline src/pages/courses/[course].tsx:53]: This basically treats the page as SSR if the route is not found in the pre-generated paths.

File: src/pages/api/revalidate.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import type { NextApiRequest, NextApiResponse } from 'next';
+
+import rateLimit, { getIP, MAX_POSTS_PER_PERIOD } from 'lib/rate-limit';
+
+const limiter = rateLimit({
```
Comment (ajhenri) [inline src/pages/api/revalidate.ts:5]: Nice! I should probably add rate limiting to the DevCenter endpoint as well.

File: src/pages/api/revalidate.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  if (req.method !== 'POST') {
+    return res.status(405).send('This is a POST-only endpoint');
+  }
+
+  try {
+    const IP = getIP(req);
+    await limiter.check(res, MAX_POSTS_PER_PERIOD, IP || '');
+  } catch {
+    return res.status(500).json({ error: { message: 'Something went wrong' } });
```
Comment (ajhenri) [inline src/pages/api/revalidate.ts:30]: Nit: shouldn't this return a `429`?

File: src/pages/api/revalidate.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  try {
+    await res.revalidate('/academia');
+    if (contentType !== 'Course') {
+      return res.json({ revalidated: true });
+    }
+    await res.revalidate(`/academia/courses/${slug}`);
+  } catch (err) {
+    return res.status(500).send('Error revalidating');
```
Comment (ajhenri) [inline src/pages/api/revalidate.ts:48]: Would be helpful to have a console.error here too

**Review by ajhenri — APPROVED** [review]

File: src/pages/api/revalidate.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import type { NextApiRequest, NextApiResponse } from 'next';
+
+import rateLimit, { getIP, MAX_POSTS_PER_PERIOD } from 'lib/rate-limit';
+
+const limiter = rateLimit({
+  max: 5, // cache limit of 5 per 60 second period.
+  ttl: 60 * 1000,
+});
+
+const reavlidateHandler = async (req: NextApiRequest, res: NextApiResponse) => {
```
Comment (ajhenri) [inline src/pages/api/revalidate.ts:10]: nit; typo

**Review by ajhenri — APPROVED** [review]

File: src/pages/api/revalidate.ts (hunk 4 above)
Comment (mcolella14) [reply src/pages/api/revalidate.ts:30]: Ah good call!

File: src/pages/courses/[course].tsx [hunk 7]
```diff
@@ -24,11 +24,7 @@ export default function CoursePage({ openForm, content }: CoursePageProps) {
       <main sx={styles.CoursePageMain}>
         {/* @ts-ignore */}
         <GridLayout sx={styles.CoursePageGrid}>
-          <CourseBody
-            formatText={mockFormatText}
-            outlineText={mockOutlineText}
-            wrapperStyles={styles.CoursePageBody}
```
Comment (justinmckenzie) [inline src/pages/courses/[course].tsx:30]: might just want to double check removing these styles are okay, looks like it just has an `order` on it

Comment (mcolella14) [reply src/pages/courses/[course].tsx:30]: Ah good call, didn't mean to remove the styles.

**Review by justinmckenzie — APPROVED** [review]

**Review by gfunk1230 — APPROVED** [review]

**Pull request merged**